libcache/cache.c

```diff
@@ -0,0 +1,649 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Cache library
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Malgorzata Wrobel
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "cache.h"
+
+#include <sys/list.h>
+// #include "list.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+
+
+#define LIBCACHE_NUM_WAYS   4
+#define LIBCACHE_ADDR_WIDTH 64
+
+
+#define IS_VALID(f) (((f) & (1 << 0)) != 0 ? 1 : 0)
+#define IS_DIRTY(f) (((f) & (1 << 1)) != 0 ? 1 : 0)
+
+#define SET_VALID(f) \
+	do { \
+		(f) |= (1 << 0); \
+	} while (0)
+#define CLEAR_VALID(f) \
+	do { \
+		(f) &= ~(1 << 0); \
+	} while (0)
+
+#define SET_DIRTY(f) \
+	do { \
+		(f) |= (1 << 1); \
+	} while (0)
+#define CLEAR_DIRTY(f) \
+	do { \
+		(f) &= ~(1 << 1); \
+	} while (0)
+
+
+#define LOG2(x) ((uint8_t)(8 * sizeof(unsigned long) - __builtin_clzl((x)) - 1))
+
+
+typedef struct cacheline_s cacheline_t;
+
+
+struct cacheline_s {
+	uint64_t tag;
+	void *data;
+	uint64_t offset;
+	unsigned char flags;
+	cacheline_t *prev, *next; /* Circular doubly linked list */
+};
+
+
+typedef struct cacheset_s {
+	cacheline_t *timestamps;
+	cacheline_t *tags[LIBCACHE_NUM_WAYS];
+	cacheline_t lines[LIBCACHE_NUM_WAYS];
+	size_t count;
+} cacheset_t;
```

Comment:
typedefed struct should be anonymous when possible
```suggestion
typedef struct {
	cacheline_t *timestamps;
	cacheline_t *tags[LIBCACHE_NUM_WAYS];
	cacheline_t lines[LIBCACHE_NUM_WAYS];
	size_t count;
} cacheset_t;
```